src/main/java/org/opentripplanner/updater/trip/AddedStopTime.java

```diff
@@ -13,41 +15,105 @@
  */
 final class AddedStopTime {
 
-  @Nullable
-  private final PickDrop pickup;
+  private final GtfsRealtime.TripUpdate.StopTimeUpdate stopTimeUpdate;
 
-  @Nullable
-  private final PickDrop dropOff;
-
-  public static final PickDrop DEFAULT_PICK_DROP = PickDrop.SCHEDULED;
-
-  AddedStopTime(@Nullable PickDrop pickup, @Nullable PickDrop dropOff) {
-    this.pickup = pickup;
-    this.dropOff = dropOff;
+  AddedStopTime(GtfsRealtime.TripUpdate.StopTimeUpdate stopTimeUpdate) {
+    this.stopTimeUpdate = stopTimeUpdate;
   }
 
   PickDrop pickup() {
-    return Objects.requireNonNullElse(pickup, DEFAULT_PICK_DROP);
+    return getPickDrop(
+      getStopTimePropertiesExtension()
+        .map(MfdzRealtimeExtensions.StopTimePropertiesExtension::getPickupType)
+        .orElse(null)
+    );
   }
 
   PickDrop dropOff() {
-    return Objects.requireNonNullElse(dropOff, DEFAULT_PICK_DROP);
+    return getPickDrop(
+      getStopTimePropertiesExtension()
+        .map(MfdzRealtimeExtensions.StopTimePropertiesExtension::getDropoffType)
+        .orElse(null)
+    );
   }
 
-  static AddedStopTime ofStopTime(GtfsRealtime.TripUpdate.StopTimeUpdate props) {
-    if (props.getStopTimeProperties().hasExtension(MfdzRealtimeExtensions.stopTimeProperties)) {
-      var ext = props
-        .getStopTimeProperties()
-        .getExtension(MfdzRealtimeExtensions.stopTimeProperties);
-      var pickup = ext.getPickupType();
-      var dropOff = ext.getDropoffType();
-      var dropOffType = PickDropMapper.map(dropOff.getNumber());
-      var pickupType = PickDropMapper.map(pickup.getNumber());
-      return new AddedStopTime(pickupType, dropOffType);
-    } else {
-      var pickDrop = toPickDrop(props.getScheduleRelationship());
-      return new AddedStopTime(pickDrop, pickDrop);
+  private PickDrop getPickDrop(
+    @Nullable MfdzRealtimeExtensions.StopTimePropertiesExtension.DropOffPickupType extensionDropOffPickup
+  ) {
+    if (isSkipped()) {
+      return PickDrop.CANCELLED;
     }
+
+    if (extensionDropOffPickup == null) {
+      return toPickDrop(stopTimeUpdate.getScheduleRelationship());
+    }
+
+    return PickDropMapper.map(extensionDropOffPickup.getNumber());
+  }
+
+  private Optional<MfdzRealtimeExtensions.StopTimePropertiesExtension> getStopTimePropertiesExtension() {
+    return stopTimeUpdate
+        .getStopTimeProperties()
+        .hasExtension(MfdzRealtimeExtensions.stopTimeProperties)
+      ? Optional.of(
+        stopTimeUpdate
+          .getStopTimeProperties()
+          .getExtension(MfdzRealtimeExtensions.stopTimeProperties)
+      )
+      : Optional.empty();
+  }
+
+  OptionalLong getArrivalTime() {
```

Comment:
I know that this is anything but consistent in the code base at the moment, but we are not using the `getXXX` style of getters in new code anymore.

https://github.com/opentripplanner/OpenTripPlanner/blob/7ff692cad27f02a77c0fd042ed4c0dfa5b9b54d9/doc/dev/decisionrecords/NamingConventions.md#L41

This class contains methods like `pickup` which use the new style, so please also rename the public ones that you added.